### PR TITLE
N3 6 add refresh token to jwt authentication

### PR DESCRIPTION
**Pull request by dmkofgl**

### CODE REVIEW

File: src/main/java/dl/news/portal/domain/entity/RefreshToken.java [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+    public String getAuthorities() {
+        return authorities;
+    }
+
+    public void setAuthorities(String authorities) {
+        this.authorities = authorities;
+    }
+
+    @Override
+    public String toString() {
```
Comment (avinahradau) [inline src/main/java/dl/news/portal/domain/entity/RefreshToken.java]: Do you really need this method?

Comment (avinahradau) [reply src/main/java/dl/news/portal/domain/entity/RefreshToken.java]: and why its before getter and setter?

File: src/main/java/dl/news/portal/domain/service/impl/RefreshTokenServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+
+import javax.persistence.EntityNotFoundException;
+import java.util.Date;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+@Service
+public class RefreshTokenServiceImpl implements RefreshTokenService {
+
+    private static final long REFRESH_TOKEN_VALIDITY_SECONDS = 5 * 60 * 60 * 24;
```
Comment (avinahradau) [inline src/main/java/dl/news/portal/domain/service/impl/RefreshTokenServiceImpl.java]: What does this numbers means?

File: src/main/java/dl/news/portal/domain/service/impl/RefreshTokenServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+    @Override
+    public RefreshToken createRefreshToken(Authentication authentication) {
+        final String authorities = authentication.getAuthorities().stream()
+                .map(GrantedAuthority::getAuthority)
+                .collect(Collectors.joining(","));
+        RefreshToken refreshToken = new RefreshToken();
+        User owner = userService.findUserByUsername(authentication.getName()).orElseThrow(EntityNotFoundException::new);
+        refreshToken.setOwner(owner);
+        refreshToken.setIssuedAt(new Date());
+        refreshToken.setExpiration(new Date(refreshToken.getIssuedAt().getTime() + REFRESH_TOKEN_VALIDITY_SECONDS * 1000));
```
Comment (avinahradau) [inline src/main/java/dl/news/portal/domain/service/impl/RefreshTokenServiceImpl.java]: what is 1000?

File: src/main/java/dl/news/portal/domain/service/impl/UserDetailServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
+    @Override
+    public UserDetails loadUserByUsername(String s) throws UsernameNotFoundException {
+
+        return userService.findUserByUsername(s)
+                .map(this::convertUser)
+                .orElseThrow(() -> new UsernameNotFoundException("No user with username " + s));
+    }
+
+    private UserDetails convertUser(User user) {
```
Comment (avinahradau) [inline src/main/java/dl/news/portal/domain/service/impl/UserDetailServiceImpl.java]: Convert user to what?

File: src/main/java/dl/news/portal/domain/service/impl/UserDetailServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+        return new org.springframework.security.core.userdetails.User(
+                user.getUsername(),
+                passwordEncoder.encode(user.getPassword()),
+                Collections.singletonList(new SimpleGrantedAuthority("user")));
```
Comment (avinahradau) [inline src/main/java/dl/news/portal/domain/service/impl/UserDetailServiceImpl.java]: use constant here

File: src/main/java/dl/news/portal/web/config/security/jwt/JwtAuthenticationFilter.java [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+
+import javax.servlet.FilterChain;
+import javax.servlet.ServletException;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+
+public class JwtAuthenticationFilter extends OncePerRequestFilter {
+    private static final String HEADER_STRING = "Authorization";
+    private static final String TOKEN_PREFIX = "Bearer ";
```
Comment (avinahradau) [inline src/main/java/dl/news/portal/web/config/security/jwt/JwtAuthenticationFilter.java:25]: add some space between constants and fields

**Review by avinahradau — CHANGES_REQUESTED** [review]

File: src/main/java/dl/news/portal/domain/service/impl/RefreshTokenServiceImpl.java (hunk 2 above)
Comment (dmkofgl) [reply src/main/java/dl/news/portal/domain/service/impl/RefreshTokenServiceImpl.java]: Time in seconds while refresh token is valid

File: src/main/java/dl/news/portal/domain/entity/RefreshToken.java (hunk 1 above)
Comment (dmkofgl) [reply src/main/java/dl/news/portal/domain/entity/RefreshToken.java]: No, i don't. I see.

File: src/main/java/dl/news/portal/domain/service/impl/RefreshTokenServiceImpl.java (hunk 3 above)
Comment (dmkofgl) [reply src/main/java/dl/news/portal/domain/service/impl/RefreshTokenServiceImpl.java]: It's magic digit )

File: src/main/java/dl/news/portal/web/config/security/jwt/JwtAuthenticationFilter.java (hunk 6 above)
Comment (dmkofgl) [reply src/main/java/dl/news/portal/web/config/security/jwt/JwtAuthenticationFilter.java:25]: Done

File: src/main/java/dl/news/portal/domain/service/impl/UserDetailServiceImpl.java (hunk 5 above)
Comment (dmkofgl) [reply src/main/java/dl/news/portal/domain/service/impl/UserDetailServiceImpl.java]: Ok

File: src/main/java/dl/news/portal/domain/service/impl/UserDetailServiceImpl.java (hunk 4 above)
Comment (dmkofgl) [reply src/main/java/dl/news/portal/domain/service/impl/UserDetailServiceImpl.java]: To UserDetails

File: src/main/java/dl/news/portal/domain/service/impl/RefreshTokenServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+
+import javax.persistence.EntityNotFoundException;
+import java.util.Date;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+@Service
+public class RefreshTokenServiceImpl implements RefreshTokenService {
+
+    private static final long REFRESH_TOKEN_VALIDITY_MILLISECONDS = 5 * 60 * 60 * 24 * 1000;
```
Comment (avinahradau) [inline src/main/java/dl/news/portal/domain/service/impl/RefreshTokenServiceImpl.java]: Remove calculation here

File: src/main/java/dl/news/portal/domain/service/impl/RefreshTokenServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+    }
+
+    @Override
+    public String buildToken(RefreshToken refreshToken) {
+        return Jwts.builder()
+                .setSubject(refreshToken.getId().toString())
+                .claim(AUTHORITIES_KEY, refreshToken.getAuthorities())
+                .signWith(SignatureAlgorithm.HS256, SIGNING_KEY)
+                .setIssuedAt(new Date(System.currentTimeMillis()))
+                .setExpiration(new Date(System.currentTimeMillis() + REFRESH_TOKEN_VALIDITY_MILLISECONDS * 1000))
```
Comment (avinahradau) [inline src/main/java/dl/news/portal/domain/service/impl/RefreshTokenServiceImpl.java]: Issue with 1000 still here, use some variable or constant here

File: src/main/java/dl/news/portal/web/config/security/JwtAuthenticationConfig.java [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+                .passwordEncoder(passwordEncoder);
+    }
+
+    @Override
+    protected void configure(HttpSecurity http) throws Exception {
+        http.cors().and().csrf().disable().
+                authorizeRequests()
+                .mvcMatchers(HttpMethod.POST, "/token/**").permitAll()
+                .mvcMatchers(HttpMethod.POST, "/users").permitAll()
+//                .mvcMatchers(HttpMethod.GET, "/**").permitAll()
```
Comment (avinahradau) [inline src/main/java/dl/news/portal/web/config/security/JwtAuthenticationConfig.java]: Comment?

**Review by avinahradau — CHANGES_REQUESTED** [review]

File: src/main/java/dl/news/portal/domain/service/impl/RefreshTokenServiceImpl.java (hunk 7 above)
Comment (dmkofgl) [reply src/main/java/dl/news/portal/domain/service/impl/RefreshTokenServiceImpl.java]: Ok

File: src/main/java/dl/news/portal/domain/service/impl/RefreshTokenServiceImpl.java (hunk 8 above)
Comment (dmkofgl) [reply src/main/java/dl/news/portal/domain/service/impl/RefreshTokenServiceImpl.java]: Done

File: src/main/java/dl/news/portal/web/config/security/JwtAuthenticationConfig.java (hunk 9 above)
Comment (dmkofgl) [reply src/main/java/dl/news/portal/web/config/security/JwtAuthenticationConfig.java]: No.It isn't need to use. Thank you

**Review by avinahradau — APPROVED** [review]

**Pull request merged**